KEP 4192: Add SVM beta docs (#52936)
kubernetes/website: content/en/docs/concepts/overview/working-with-objects/storage-version.md

---
title: Storage Versions
content_type: concept
weight: 110
---

<!-- overview -->

> The Kubernetes API server stores objects, relying on an etcd-compatible backing

Author: Missed a bit from https://github.com/kubernetes/website/pull/52936#discussion_r2553974944, fixed!

underlying binary encodings of the object matter, such as for encryption at
rest, or API deprecation.

The same API may have multiple storage versions that the API Server can then
convert to an object schema. A single object that is part of that resource must
only have one storage version at any time. This means that the API Server is
aware of the binary encodings of the objects and is able to convert between all
the stored versions to the API Representation of the object dynamically.

The version of an object is separate from the storage version entirely. For
example, a `v1alpha1` and `v1beta1` API Object for the same Resource will be
encoded the same in storage as long as the storage version has not been updated
between the two objects.

<!-- body -->

## Storage version to resource mapping

Every resource will have 1 active storage version at any point in time, meaning
that any write to an object will store the object at that storage version. The
storage version can be updated however, making it so that objects can be stored
at differing versions. One object will only be stored at one storage version at
any time.

> Reads from the APIServer will simply convert the stored data to the API

Reviewer: nit: to match the previously-used form
```suggestion
Reads from the API Server will simply convert the stored data to the API
```

representation of the object. This makes it so that old storage versions can sit
indefinitely as long as no updates occur to the object. Writes on the other

> hand, will convert the stored object to the new representation upon update. 

Author: Yep, that makes it flow better.

    shortNames:
    - ct
```

The `v1beta1` API definition is used as the storage version, meaning that any
updates or creation of `crontabs` will be stored with the object schema of the
`v1beta1` api. In this case it actually would mean that the `v1` API object
would never be able to store the `time` field since it is not part of the
storage definition.

Upon modification of the version that is used for storage, that version of the
API will be used to store any new or update CRs. Watching or getting the object
will have the object be in use but will just convert the object from the old
storage version and not affect the object. Only updating or creating will have
an effect and use the newly defined storage version.  

## How storage versions are relevant to encryption at rest

There are tools to [encrypt the at rest
storage](/docs/tasks/administer-cluster/kms-provider/) of a cluster, especially
for cluster secrets. This adds an additional layer of protection for data
exfiltration since the actual stored data in the cluster is encrypted. This
means that the API Server is actually decrypting the stored data when it decodes

> the data from storage. The APIServer must have the key for that storage version

Author: Thanks, finding enough ways to say store/encode etc... in this doc was hard hahaha!

at differing versions. One object will only be stored at one storage version at
any time.

Reads from the API Server will simply convert the stored data to the API
representation of the object. This makes it so that old storage versions can sit
indefinitely as long as no updates occur to the object. Writes, on the other
hand, will convert the stored object to the new representation upon update. 

## Storage versions for custom resources {#CustomResourceDefinition-storage-version}

[Custom
resources](/docs/concepts/extend-kubernetes/api-extension/custom-resources/#Storage) are
defined dynamically, and as such differ from built in Kubernetes types with
their storage version. Builtin objects generally have their storage encoding
defined separately from their API types, where the stored object acts as a hub
and the specific version of the resource does not matter apart from being a
field in the object schema. 

However, for custom resources, a certain version of the resource must be set as
the storage version. The schema defined by that specific version of the custom
resource will be used as the encoding of the resource in the storage layer. See

> [the advanced CRD

Author: Done

featureset](/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility)
for more detailed information on the API setup and versioning.

> For example see the custom resource definition crontabs.

Author: Done

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:

>   # name must match the spec fields below, and be in the form: <plural>.<group>

Reviewer 2: I'd omit this comment unless relevant to the current page.

  name: crontabs.example.com
spec:
  # group name to use for REST API: /apis/<group>/<version>
  group: example.com
  # list of versions supported by this CustomResourceDefinition
  versions:
  - name: v1beta1
    # Each version can be enabled/disabled by Served flag.
    served: true
    # One and only one version must be marked as the storage version.
    storage: true
    # A schema is required
    schema:
      openAPIV3Schema:
        type: object
        properties:
          host:
            type: string
          port:
            type: string
  - name: v1
    served: true
    storage: false
    schema:
      openAPIV3Schema:
        type: object
        properties:
          host:
            type: string
          port:
            type: string
          time:
            type: string
  # The conversion section is introduced in Kubernetes 1.13+ with a default value of
  # None conversion (strategy sub-field set to None).
  conversion:
    # None conversion assumes the same schema for all versions and only sets the apiVersion
    # field of custom resources to the proper value
    strategy: None

>   # either Namespaced or Cluster

Author: Done

  scope: Namespaced
  names:

>     # plural name to be used in the URL: /apis/<group>/<version>/<plural>

Reviewer 2: I'd omit this comment unless relevant to the current page.

    plural: crontabs
    # singular name to be used as an alias on the CLI and for display
    singular: crontab

>     # kind is normally the CamelCased singular type. Your resource manifests use this.

Reviewer 2: I'd omit this comment unless relevant to the current page.

Author: Done

    kind: CronTab

>     # shortNames allow shorter string to match your resource on the CLI

Reviewer 2: I'd omit this comment unless relevant to the current page.

Author: Done

    shortNames:
    - ct
```

> The `v1beta1` API definition is used as the storage version, meaning that any

Reviewer 2: Maybe say why (ie, that it is `stored`), and that trying to set two versions as stored is invalid.

updates or creation of `crontabs` will be stored with the object schema of the
`v1beta1` api. In this case it actually would mean that the `v1` API object
would never be able to store the `time` field since it is not part of the
storage definition.

Upon modification of the version that is used for storage, that version of the
API will be used to store any new or update CRs. Watching or getting the object
will have the object be in use but will just convert the object from the old
storage version and not affect the object. Only updating or creating will have
an effect and use the newly defined storage version.  

## How storage versions are relevant to encryption at rest

There are tools to [encrypt the at rest
storage](/docs/tasks/administer-cluster/kms-provider/) of a cluster, especially
for cluster secrets. This adds an additional layer of protection for data
exfiltration since the actual stored data in the cluster is encrypted. This
means that the API Server is actually decrypting the data as it retrieves them
from storage. the data from storage. The APIServer must have the key for that
storage version in order to decode the object properly.

> The storage version in this case is more than just the binary encoding of the

Author: I think these things are intertwined no? When secrets are encrypted at rest that means the encryption key is part of the storage version since it must be used to unencrypt the data. Unless the data is modified later via an update, the secret will use the old encryption key which makes an encryption key effectively an input into the "storage version".

---
title: Storage Versions
content_type: concept
weight: 20
---

<!-- overview -->

> Storage version is the current version that the API object is stored at in the

Reviewer 2: ```suggestion
The Kubernetes API server stores objects, relying on an etcd-compatible backing store
(often, the backing storage is etcd itself).
Each object is serialized using a particular version of that API type;
for example, the v1 representation of a ConfigMap.
Kubernetes uses the term _storage version_ to describe how an object is stored in your
```

Author: Done

> cluster. While it does not affect what the API Server returns it is an important

Reviewer 2: ```suggestion
cluster.
The Kubernetes API also relies on automatic conversion; for example, if you have a HorizontalPodAutoscaler,
then you can interact with that HorizontalPodAutoscaler using any mix of the v1 and v2 versions of the
HorizontalPodAutoscaler API.
Kubernetes is responsible for converting each API call so that clients do not see what version is actually serialized.

For cluster administrators, object storage version is an important
```

Author: Added

concept to understand since it is what links the API representation of the
object to the actual encoding in the storage backend. This can matter for when
the underlying binary encodings of the object matter, such as for encryption at
rest, or API deprecation.

The same API object may have multiple storage versions that the API Server can
then convert to an object schema. This means that the API Server is aware of the
binary encodings of the objects and is able to convert between all the stored
versions to the API Representation of the object dynamically.

The version of an object is separate from the storage version entirely. For

> example, a `v1alpha1` and `v1beta1` API Object for the same GroupResource will

Reviewer 2: This makes it look like you could `kubectl get GroupResources`.

Avoid mentioning implementation details that people won't see if they only interact via the HTTP APIs.

Author: Yep, removed GroupResource to ensure that users don't see this as an object.

be encoded the same in storage as long as the storage version has not been
updated between the two objects.

<!-- body -->

> ## Storage Version to Resource Mapping

Author: Done

Every resource will have 1 active storage version at any point in time, meaning
that any write to an object will store the object at that storage version. Reads
can occur from any storage version however, and will not update the stored
version of the object. The only way a storage version will get updated is by
modifying the object somehow. Otherwise the stored version will stay the same as
long as the object exists in the cluster. 

> ## Encryption at Rest

Reviewer 2: ```suggestion
## How storage versions are relevant to encryption at rest
```
?

> 

Author: Yeah good point, I don't think we have anything at this point. Probably something to add in 1.36?

There are tools to [encrypt the at rest
storage](/docs/tasks/administer-cluster/kms-provider/) of a cluster, especially
for cluster secrets. This adds an additional layer of protection for data
exfiltration since the actual stored data in the cluster is encrypted. This
means that the API Server is actually decrypting the stored data when it decodes
the data from storage. The APIServer must have the key for that storage version

> in order to decode the object properly.

Reviewer: The key is not actually stored per storage version, but you may want to migrate upon a key-encryption-key rotation. @enj or @aramase please correct me if this is still inaccurate.

Author: Yeah, not as familiar with encryption as with the multiple API versions, correct me if I have any misunderstandings here. I was under the impression that the key couldn't be deprecated until all objects of the resource had been updated since otherwise the object may be stored encrypted with the old key.

Reviewer: If you remove access to the key while there are some objects still encrypted with it, you lose access to those objects, that's indeed correct.

Author: Sounds good, resolving!

## Custom Resource Definition Stored Version

[Custom
resources](/docs/concepts/extend-kubernetes/api-extension/custom-resources/) are

> defined dynamically, and as such differ from built in kubernetes types with

Author: Done

their storage version. Builtin objects generally have their storage version

> defined as a seperate object from their API types, where the stored version acts

Reviewer 2: This makes it sound like you could `kubectl get storedversions/pods` or something. Avoid implying that.

as a hub and the specific version of the resource does not matter apart from
being a field in the object schema. This is why for custom resources, a certain
version of the resource must be set as the storage version. That object encoding

> will be used as the schema for that storage version. See [versioning custom

Reviewer 2: Try to hyperlinks to concept pages. If the CRD concept docs don't have anything, not even a stub heading about CRD storage versions, add that stub doc to the other concept page, and link there.

resources](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/)
for more detailed information on the API setup.

Upon modification of the version that is used for storage, that version of the
API will be used to store any new or update CRs. The old version will still be
used for any reads, but new writes will use the updated stored version.

> # Migrating Storage Versions

Reviewer 2: At this point, we're talking about a task and can (should!) link to the doc we already have.

Multiple storage versions can pose problems for cluster administrators. Since a

> resource must be actively in use to update the storage version, when a key

Reviewer: I'd probably say that watching/getting a resource classifies as it being actively in use but not having an effect on its storage version :thinking:

Author: Yeah makes more sense that way, added to the doc.

rotation is done, both the old encryption key and the new encryption key must

> remain in use until the administrator is sure all objects have been written to

Author: Moved this below the simpler versioning issue to build from simpler to bigger concepts. This is an important use case for migration however since we can't deprecate old encryption keys until this migration occurs though.

at least once. This poses both security risks and usability issues, since a key
cannot be fully removed from use until then. 

> Another important issue is the deprecation of old storage versions, such as with

Reviewer: Should this rather appear in the section above? Should we structure this whole page differently so that maybe we describe what storage version migration is and then when it would be useful?

Looking at what we've got now, this is a page about Storage Versions, but one of the first sections is about "Encryption at rest", which actually has very little to do with storage versions.

CRDs. A cluster administrator may not remove old versions of the API which may
be unsupported until they are sure that all objects have been moved off of it.
With a large number of objects and an opaque view into which ones are new and
which ones still are backed by old storage versions, it makes it difficult to
tell when a version can be safely removed. If a version is removed prematurely,
it can mean being unable to read the object entirely.

> With `StorageVersionMigration` a cluster administrator can kick off a migration

Reviewer 2: This reads like an announcement; avoid that. We aim for [timeless](https://developers.google.com/style/timeless-documentation) documentation, within reason.

---
title: Storage Versions
content_type: concept

> weight: 20

Reviewer 2: Set it to a multiple of 10 larger than all its siblings are using.
